### ISSUE TITLE
union together with azureblob causes SIGSEGV panic  

### ISSUE THREAD

**Issue by jeff1985** [rclone/rclone#4851]
#### What is the problem you are having with rclone?

I tried merging two azure SAS endpoints via union storage system. When I click on a subpath in the http serving, the process fails and causes SIGSEGV panic.  

If I use the individual azureblob storage system for running serve, it works fine. 

#### What is your rclone version (output from `rclone version`)

```
rclone v1.53.3
- os/arch: darwin/amd64
- go version: go1.15.5
```

#### Config ~/.config/rclone/rclone.conf

```
[azuresas1]
type = azureblob
sas_url = https://xxxxx.blob.core.windows.net/mycontainer/?SAS

[azuresas2]
type = azureblob
sas_url = https://xxxxx2.blob.core.windows.net/mycontainer/?SAS

[merged]
type = union
upstreams = azuresas1::ro azuresas2::ro
action_policy = epall
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

MacOSX 10.15.7

####  Which cloud storage system are you using? (e.g. Google Drive)

Azureblob

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone serve http merged:

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2020/12/07 16:32:46 DEBUG : rclone: Version "v1.53.3" starting with parameters ["rclone" "serve" "http" "merged:" "-vv"]
2020/12/07 16:32:46 DEBUG : Using config file from "/Users/jeff/.config/rclone/rclone.conf"
2020/12/07 16:32:46 DEBUG : Creating backend with remote "merged:"
2020/12/07 16:32:46 DEBUG : Creating backend with remote "azuresas1:"
2020/12/07 16:32:46 DEBUG : Creating backend with remote "azuresas2:"
2020/12/07 16:32:46 DEBUG : union root '': actionPolicy = *policy.EpAll, createPolicy = *policy.EpMfs, searchPolicy = *policy.FF
2020/12/07 16:32:46 INFO  : union root '': poll-interval is not supported by this remote
2020/12/07 16:32:46 NOTICE: union root '': Serving on http://127.0.0.1:8080/
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4617b3b]

goroutine 102 [running]:
github.com/Azure/azure-storage-blob-go/azblob.containerClient.ListBlobHierarchySegment(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000506560, 0xc, 0x0, ...)
	github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zz_generated_container.go:683 +0x49b
github.com/Azure/azure-storage-blob-go/azblob.ContainerURL.ListBlobsHierarchySegment(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000506560, 0xc, 0x0, ...)
	github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/url_container.go:249 +0x271
github.com/rclone/rclone/backend/azureblob.(*Fs).list.func1(0xc00010e118, 0x8, 0x0)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:623 +0x174
github.com/rclone/rclone/fs.pacerInvoker(0x1, 0xa, 0xc000212150, 0x68, 0x8203748, 0xc000212150)
	github.com/rclone/rclone/fs/fs.go:1477 +0x3c
github.com/rclone/rclone/lib/pacer.(*Pacer).call(0xc0000a63c0, 0xc000212150, 0xa, 0x70, 0x55d52a0)
	github.com/rclone/rclone/lib/pacer/pacer.go:197 +0x90
github.com/rclone/rclone/lib/pacer.(*Pacer).Call(0xc0000a63c0, 0xc000212150, 0xb, 0xc0004f7000)
	github.com/rclone/rclone/lib/pacer/pacer.go:216 +0x6a
github.com/rclone/rclone/backend/azureblob.(*Fs).list(0xc0001ac000, 0x598d180, 0xc00019c018, 0xc0001762f8, 0xb, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:621 +0x885
github.com/rclone/rclone/backend/azureblob.(*Fs).listDir(0xc0001ac000, 0x598d180, 0xc00019c018, 0xc0001762f8, 0xb, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:697 +0x116
github.com/rclone/rclone/backend/azureblob.(*Fs).List(0xc0001ac000, 0x598d180, 0xc00019c018, 0xc0001762f8, 0xb, 0x406f07e, 0xc00059a788, 0xc00019c000, 0x0, 0xc0007d3768)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:755 +0x16a
github.com/rclone/rclone/backend/union.(*Fs).List.func1(0x0)
	github.com/rclone/rclone/backend/union/union.go:579 +0xa7
github.com/rclone/rclone/backend/union.multithread.func1(0xc00059a790, 0xc0004239a0, 0x0)
	github.com/rclone/rclone/backend/union/union.go:881 +0x58
created by github.com/rclone/rclone/backend/union.multithread
	github.com/rclone/rclone/backend/union/union.go:879 +0x8c
```

**Comment by ncw** [comment]
Looking through the call chain I see this which has a nil `ctx` variable

```
github.com/Azure/azure-storage-blob-go/azblob.ContainerURL.ListBlobsHierarchySegment(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000506560, 0xc, 0x0, ...)
	github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/url_container.go:249 +0x271
```

However it was called from here where it looks fine

```
github.com/rclone/rclone/backend/azureblob.(*Fs).list(0xc0001ac000, 0x598d180, 0xc00019c018, 0xc0001762f8, 0xb, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	github.com/rclone/rclone/backend/azureblob/azureblob.go:621 +0x885
```

I don't see how this could have happened - the ctx variable isn't written between those calls. So I suspect some kind of memory scribble which is very unusual with Go programs.

Maybe it could be
- a problem with the Go runtime / compiler
- a problem with the RAM on your machine
- a problem with a bit of code using CGO (there shouldn't be any code using CGO in the configuration you are using)

Can you reproduce this reliably?

Is it always the same sequence of functions in the call trace?

This could be a problem with RAM in your machine - is it generally reliable? Have you run a RAM tester on it?

Can you try with [the latest beta](https://beta.rclone.org) which has a newer Azure blob library and was compiled with a newer go compiler and runtime.

Thanks

**Comment by jeff1985** [comment]
Hi @ncw 

I'm running rclone locally on a high end MacBook which is only 6 months old and I did not notice any RAM problems while using other software.

I be sure, I have executed latest beta via docker on a cloud virtual machine which is running on Linux and fully separated from my local environment. The result is the same. 

```
root@xxx:~/rclonetest# docker run -p 0.0.0.0:8088:8080/tcp -v /root/rclonetest/rclone.conf:/config/rclone/rclone.conf rclone/rclone:beta serve http merged:/ --addr 0.0.0.0:8080 -vv
2020/12/09 21:45:55 DEBUG : rclone: Version "v1.54.0-beta.4965.cb16f4207" starting with parameters ["rclone" "serve" "http" "merged:/" "--addr" "0.0.0.0:8080" "-vv"]
2020/12/09 21:45:55 DEBUG : Using config file from "/config/rclone/rclone.conf"
2020/12/09 21:45:55 DEBUG : Creating backend with remote "merged:/"
2020/12/09 21:45:55 DEBUG : Creating backend with remote "azuresas2:"
2020/12/09 21:45:55 DEBUG : Creating backend with remote "azuresas1:"
2020/12/09 21:45:55 DEBUG : union root '/': actionPolicy = *policy.EpAll, createPolicy = *policy.EpMfs, searchPolicy = *policy.FF
2020/12/09 21:45:55 INFO  : union root '/': poll-interval is not supported by this remote
2020/12/09 21:45:55 NOTICE: union root '/': Serving on http://[::]:8080/
2020/12/09 21:46:14 INFO  : : 78.54.141.187:56335: Serving directory
2020/12/09 21:46:14 INFO  : favicon.ico: 78.54.141.187:56335: File not found
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xa2dc7b]

goroutine 29 [running]:
github.com/Azure/azure-storage-blob-go/azblob.containerClient.ListBlobHierarchySegment(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00060d760, 0xc, 0x0, ...)
	/go/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.11.0/azblob/zz_generated_container.go:693 +0x49b
github.com/Azure/azure-storage-blob-go/azblob.ContainerURL.ListBlobsHierarchySegment(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00060d760, 0xc, 0x0, ...)
	/go/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.11.0/azblob/url_container.go:251 +0x19b
github.com/rclone/rclone/backend/azureblob.(*Fs).list.func1(0xc000068050, 0xc000068050, 0x203000)
	/go/src/github.com/rclone/rclone/backend/azureblob/azureblob.go:707 +0x174
github.com/rclone/rclone/fs.pacerInvoker(0x1, 0xa, 0xc000571960, 0x68, 0x7f6eeedaa108, 0xc000571960)
	/go/src/github.com/rclone/rclone/fs/fs.go:1505 +0x3c
github.com/rclone/rclone/lib/pacer.(*Pacer).call(0xc0004cd440, 0xc000571960, 0xa, 0x70, 0x1a89d80)
	/go/src/github.com/rclone/rclone/lib/pacer/pacer.go:197 +0x90
github.com/rclone/rclone/lib/pacer.(*Pacer).Call(0xc0004cd440, 0xc000571960, 0xb, 0x0)
	/go/src/github.com/rclone/rclone/lib/pacer/pacer.go:216 +0x6a
github.com/rclone/rclone/backend/azureblob.(*Fs).list(0xc000558a00, 0x1e63e20, 0xc000042020, 0xc00017a2f8, 0xb, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	/go/src/github.com/rclone/rclone/backend/azureblob/azureblob.go:705 +0x866
github.com/rclone/rclone/backend/azureblob.(*Fs).listDir(0xc000558a00, 0x1e63e20, 0xc000042020, 0xc00017a2f8, 0xb, 0x0, 0x0, 0x0, 0x0, 0x1, ...)
	/go/src/github.com/rclone/rclone/backend/azureblob/azureblob.go:781 +0x119
github.com/rclone/rclone/backend/azureblob.(*Fs).List(0xc000558a00, 0x1e63e20, 0xc000042020, 0xc00017a2f8, 0xb, 0x46c31e, 0xc00060d478, 0xc000042000, 0x0, 0xc00004f768)
	/go/src/github.com/rclone/rclone/backend/azureblob/azureblob.go:839 +0x16a
github.com/rclone/rclone/backend/union.(*Fs).List.func1(0x0)
	/go/src/github.com/rclone/rclone/backend/union/union.go:579 +0xa7
github.com/rclone/rclone/backend/union.multithread.func1(0xc00060d620, 0xc000076280, 0x0)
	/go/src/github.com/rclone/rclone/backend/union/union.go:895 +0x58
created by github.com/rclone/rclone/backend/union.multithread
	/go/src/github.com/rclone/rclone/backend/union/union.go:893 +0x8c
```
  
This is perfectly reproducible and happens every time when I try to open a sub path from the merged filesystem.

**Comment by ncw** [comment]
> This is perfectly reproducible and happens every time when I try to open a sub path from the merged filesystem. I also tried running rclone beta locally via docker and it also failed.

I must have mis-interpreted the traceback :blush:

I managed to reproduce this with two SAS URLs

I think the problem is that the union is trying to read from the sas URL outside of the container it is for which is crashing the azureblob SDK

I fixed that here - please give it a test

[v1.54.0-beta.4967.446da5e3d.fix-4851-azure-sas](https://beta.rclone.org/branch/fix-4851-azure-sas/v1.54.0-beta.4967.446da5e3d.fix-4851-azure-sas/) on branch [fix-4851-azure-sas](https://github.com/rclone/rclone/tree/fix-4851-azure-sas) (uploaded in 15-30 mins)

This probably deserves a bug report on the Azure Blob SDK too - will see if I have the energy tomorrow!

**Comment by jeff1985** [comment]
Hi @ncw 

thanks for the fix! I tested it and it works just fine!
Please merge ;)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

Thanks for testing :-)
